---
title: 'Finding negative selection in cancer using pN/pS: insights from computer simulations'
date: '`r format(Sys.time(), "%d/%m/%Y")`'
author: "Andrés Pérez-Figueroa & David Posada"
bibliography: ["pNpS.Sims.bib"]
biblio-style: "apalike"
link-citations: true
output:
  html_notebook: 
    fig_caption: yes
    fig_height: 7
    fig_width: 10
    highlight: tango
    theme: readable
  bookdown::html_document2:
    number_sections: FALSE
    fig_caption: TRUE
    toc: false
  bookdown::pdf_document:
    base_format: rticles::elsevier_article
  html_document: 
    fig_caption: yes
    fig_height: 7
    fig_width: 10
    highlight: tango
    theme: readable
---
```{r setup, include=FALSE, message=FALSE}
library(tidyverse, quietly = T, warn.conflicts = FALSE)
library(data.table)
library(extrafont)
library(gridExtra)
library(latex2exp)
knitr::opts_chunk$set(message = FALSE, warning=FALSE, cache = FALSE,
                      fig.path="figures/", dev=c('png', 'pdf'),
                      echo = FALSE, fig.height = 12, fig.width = 16)
knitr::opts_knit$set(eval.after = "fig.cap")
caption <- "generic caption"
```



```{r load_data, include=FALSE}
csize <- 100 #max cohort size
files <- list.files(path = here::here("results"), pattern = "McF.*\\.result\\.csv", full.names = T)
big_table <- lapply(files, read_csv) %>% bind_rows() %>% filter(!is.na(s_pos), s_neg>=-1)
```


[...]     


```{r make-prog-data}
prog_data <- big_table %>%
  filter(s_pos>0) %>%
  group_by(s_pos,s_neg) %>% 
  mutate(cohort=rep(1:n(),each=csize)[1:n()]) %>% 
  group_by(cohort,add = T) %>%
  group_by(s_pos,s_neg) %>% 
  select(s_pos, s_neg, Time, tumor_size, n_clones, largest_clone, w_total) %>%
  summarise_all(mean) 
  
```

```{r fig-prog, fig.height=8, fig.width=16, fig.cap=caption}

fp1 <-gather(prog_data, "parameter", "value", Time) %>% mutate(value=log10(value)) %>%
  ggplot(aes(x=s_neg,y=value)) +
    geom_line(aes(color=as.factor(s_pos)),lwd=1.3) +
    geom_point(aes(color=as.factor(s_pos))) +
    ggpubr::theme_pubr(base_size = 20)+
    theme(axis.text.x = element_text(angle = 45, vjust = 0.47)) +
    theme(legend.position = "bottom") +
    labs(y="logartimic time", x="s-", color="      s+", subtitle = "Progression time")
fp2 <-gather(prog_data, "parameter", "value", n_clones) %>%
  ggplot(aes(x=s_neg,y=value)) +
    geom_line(aes(color=as.factor(s_pos)),lwd=1.3) +
    geom_point(aes(color=as.factor(s_pos))) +
    ggpubr::theme_pubr(base_size = 20)+
    theme(axis.text.x = element_text(angle = 45, vjust = 0.47)) +
    theme(legend.position = "none") +
    ylim(c(0,5)) +
    labs(y="number of clones", x="s-", color="      s+", subtitle = "Number of clones")
fp3 <-gather(prog_data, "parameter", "value", largest_clone) %>%
  ggplot(aes(x=s_neg,y=value)) +
    geom_line(aes(color=as.factor(s_pos)),lwd=1.3) +
    geom_point(aes(color=as.factor(s_pos))) +
    ggpubr::theme_pubr(base_size = 20)+
    theme(axis.text.x = element_text(angle = 45, vjust = 0.47)) +
    theme(legend.position = "none") +
    ylim(c(0,1)) +
    labs(y="proportion of tumor size", x="s-", color="      s+", subtitle = "Largest clone")


ggpubr::ggarrange(fp1,fp2,fp3, nrow=1, legend = "bottom", common.legend = TRUE)

caption <- "Parameters obtained at the end of the tumor progression for the different scenarios. Growth time represents the time, in decimal logarithmic scale, that the tumor took to reach the size of $10^7$ cells. The number of clones indicates the final number of clones present in the tumor, considering as clones the different genotypes with positive mutations present. The largest clone shows the proportion of the size of the tumor that occupies the majority clone."
```




[...]


```{r create-pnps_table, fig.height=6.5, fig.width=14, fig.cap=caption}
pnps_table <-big_table %>% 
  filter(s_pos>0) %>%
  group_by(s_pos,s_neg) %>% 
  summarize_at(1:18,sum) %>%
  mutate(all_driver=N_driver/S_driver,all_neutral=N_neutral/S_neutral,all_deletereous=N_deletereous/S_deletereous) %>%
  mutate(sp_driver=sp_N_driver/sp_S_driver,sp_neutral=sp_N_neutral/sp_S_neutral,sp_deletereous=sp_N_deletereous/sp_S_deletereous) %>%
  mutate(cl_driver=cl_N_driver/cl_S_driver,cl_neutral=cl_N_neutral/cl_S_neutral,cl_deletereous=cl_N_deletereous/cl_S_deletereous) %>%
  gather (gene,pN_pS,all_driver:cl_deletereous, factor_key = TRUE) %>%
  separate(gene,c("mutations","gene"), sep = "_") %>%
  select(s_pos,s_neg,mutations,gene,pN_pS) 

```

```{r fig-pnps, fig.height=15, fig.width=14, fig.cap=caption}
ind_plot <- function(X,y1=0,y2=2, tit){
  X %>%
    ggplot(aes(x=s_neg,y=pN_pS))+
      geom_line(aes(color=as.factor(s_pos)),lwd=1.3) +
      geom_point(aes(color=as.factor(s_pos))) +
      ggpubr::theme_pubr(base_size = 20)+
      geom_hline(yintercept = 1, lty="dashed") +
      labs(y="pN/pS", x="s-", color="      s+", title=tit) +
     ylim(c(y1,y2))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
      theme(plot.title = element_text(face = "bold", size=14))
  
  
}

tit.lab=c("A. Driver gene, q > 0",
                           "B. Neutral gene, q > 0",
                           "C. Deletereous gene, q > 0",
                           "D. Driver gene, q > 0.05",
                           "E. Neutral gene, q > 0.05",
                           "F. Deletereous gene, q > 0.05",
                           "G. Driver gene, q = 1",
                           "H. Neutral gene, q = 1",
                           "I. Deletereous gene, q = 1")

f1 <- pnps_table %>% filter(mutations=="all",gene=="driver") %>% ind_plot(y1=0,y2=1.5, tit.lab[1])
f2 <- pnps_table %>% filter(mutations=="all",gene=="neutral") %>% ind_plot(y1=0,y2=1.5, tit.lab[2])
f3 <- pnps_table %>% filter(mutations=="all",gene=="deletereous") %>% ind_plot(y1=0,y2=1.5, tit.lab[3])
f4 <- pnps_table %>% filter(mutations=="sp",gene=="driver") %>% ind_plot(y1=0,y2=50, tit.lab[4])
f5 <- pnps_table %>% filter(mutations=="sp",gene=="neutral") %>% ind_plot(y1=0,y2=1.5, tit.lab[5])
f6 <- pnps_table %>% filter(mutations=="sp",gene=="deletereous") %>% ind_plot(y1=0,y2=1.5, tit.lab[6])
f7 <- pnps_table %>% filter(mutations=="cl",gene=="driver") %>% ind_plot(y1=0,y2=50, tit.lab[7])
f8 <- pnps_table %>% filter(mutations=="cl",gene=="neutral") %>% ind_plot(y1=0,y2=1.5, tit.lab[8])
f9 <- pnps_table %>% filter(mutations=="cl",gene=="deletereous") %>% ind_plot(y1=0,y2=1.5, tit.lab[9])


ggpubr::ggarrange(f1,f2,f3, f4, f5,f6,f7,f8,f9, 
                  nrow = 3, ncol=3, common.legend = TRUE, heights = c(4,4,4)
                 )

caption <- "Average pN/pS, considering all the mutations present in the tumor, for each of the three genes (driver, neutral and deletereous) in all the scenarios simulated for the different positive ($s^+$, color lines) and negative ($s^-$, x-axis) selection coefficient. A.B,C show the pN/pS calculated using all the simulations present in the tumot. D, E, F using only those mutations above the threshold frequency of 0.05. G, H, I use only the clonal mutations for the tumor, those present in all the cells.
"
```



 
## VAF

```{r load-freqs}
freqsQ <- fread(here::here("results","s_freqs_merged.csv"))
```

```{r}
# Temp.
files2 <- list.files(path = here::here("results"), pattern = "McF_31\\.freqs\\.csv", full.names = T)
freqs <- lapply(files2, read_csv, col_names=FALSE) %>% bind_rows()
```

```{r}
q <- freqs[1:100,1:3000] %>% as.matrix %>% as.numeric
nq <- 100
tFreqs <- data.frame(q=q, 
                      gene=factor(c(rep("driver",nq*1000),rep("deletereous",nq*1000),rep("neutral",nq*1000)), 
                                 levels = c("driver","neutral","deletereous")),
                      mutation=factor(c(rep("N",nq*500),rep("S",nq*500),rep("N",nq*500),rep("S",nq*500),rep("N",nq*500),rep("S",nq*500)), 
                                 levels = c("N","S")))
```

```{r}
examplePNPS <- as.data.table(tFreqs)[q>0,.N, by=c("gene","mutation")] %>% 
                              dcast(gene ~ mutation) %>% 
                              mutate(pNpSall=N/S) %>% 
                              select(gene,pNpSall) %>%
                left_join( as.data.table(tFreqs)[q>0.05,.N, by=c("gene","mutation")] %>%
                             dcast(gene ~ mutation) %>% 
                             mutate(pNpSdet=N/S) %>% 
                             select(gene,pNpSdet) ) %>%
                left_join( as.data.table(tFreqs)[q==1,.N, by=c("gene","mutation")] %>%
                             dcast(gene ~ mutation) %>% 
                             mutate(pNpSclon=N/S) %>% 
                             select(gene,pNpSclon) ) 
examplePNPS
```


```{r, fig.height=6, fig.width=12, fig.cap = caption}
tFreqs %>% 
  filter(q>0) %>% 
  ggplot(aes(x=q, fill=mutation)) + 
    #geom_histogram(bins=11) +
  geom_bar(position = "stack") +
    scale_x_binned(breaks = seq(0.05,0.95, by = 0.05),
                   labels = c(".05","",".15","",".25","",".35","",".45","",".55","",".65","",".75","",".85","",".95"))+
    
    facet_wrap(~gene) +
  
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=30000, fill=NULL, 
              label=TeX('$pN/pS_{\\lbrack 0,1\\rbrack}$ = ', output = "character")),
              vjust = 1, hjust=1, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=30000, fill=NULL, 
              label=sprintf(" %.02f",pNpSall)),
              vjust = 1, hjust=0, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=26000, fill=NULL, 
              label=TeX('$pN/pS_{\\lbrack 0.05,1\\rbrack}$ = ', output = "character")),
              vjust = 1, hjust=1, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=26000, fill=NULL, 
              label=sprintf(" %.02f",pNpSdet)),
              vjust = 1, hjust=0, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=22000, fill=NULL, 
              label=TeX('$pN/pS_{\\lbrack 1,1\\rbrack}$ = ', output = "character")),
              vjust = 1, hjust=1, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=22000, fill=NULL, 
              label=sprintf(" %.02f",pNpSclon)),
              vjust = 1, hjust=0, size=5, parse=T) +
  
    labs(x="Frequency", y="ocurrences in cohort", color="mutation type", title=latex2exp::latex2exp("$s^+ = 0.1$, $s^- = -0.25$, 1 cohort 100 tumors")) +
    ggpubr::theme_pubr(base_size = 20) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
caption <- "Variant allele in a cohort of 1000 tumors for the $s^+$ = 0.1 and $s^-$ = -0.25 scenario."
```


  
## Relationship between selection coefficient and pN/pS


```{r}
freqs <- fread(here::here("results","s_freqs_merged.csv"))
#Limit to 3000 replicates,as some cases could have a couple more of them 
freqs <- freqs[, head(.SD,3000), by=V1002]
colnames(freqs) <- c("s", rep("N",500), rep("S",500),"id", rep("kk",4))
freqs <- data.table::melt(freqs, id.vars = "s", measure.vars = 2:1001, variable.name = "mutation", value.name = "q")

```

```{r}
t_pNpS_s <- freqs[(s>-1),.(m0=sum(q>0),m01=sum(q>=0.001),m5=sum(q>=0.05),m100=sum(q==1)),by=.(s,mutation)] %>% 
  dcast(s~mutation, value.var=c("m0","m01","m5","m100")) %>%
  .[,.(s, pNpS_0 = m0_N/m0_S, pNpS_0.001 = m01_N/m01_S, pNpS_0.05 = m5_N/m5_S, pNpS_1 = m100_N/m100_S)] %>%
  data.table::melt(id.vars = "s", measure = patterns("^pNpS_"), variable.name="mutations", value.name="pN_pS") 
setattr(t_pNpS_s[["mutations"]],"levels",c("all", "q >= 0.001", "q >= 0.05", "clonal(q = 1)"))
```


```{r fig.height=6, fig.width=12}
 p3 <- t_pNpS_s %>% 
  filter(mutations!="q >= 0.001") %>%
  ggplot(aes(x=s, y=pN_pS, color=mutations))+
    #geom_smooth(method = "loess", lwd=2, se=FALSE)+
    geom_line(lwd=1.2) +
    geom_point() +
     ggpubr::theme_pubr(base_size=24)+
    geom_hline(yintercept = 1, lty="dashed") +
    geom_vline(xintercept = 0, lty="dashed") +
    scale_color_discrete(name = "", labels = unname(TeX(c("$pN/pS_{\\[0,1\\]}$", 
                                               "$pN/pS_{\\[0.05,1\\]}$",
                                               "$pN/pS_{\\[1,1\\]"))))+
    labs(x="s", y="pN/pS") +
    xlim(c(-1,2)) 

p3
```


```{r}
facet.labs <- c("s = -0.75","s = -0.05","s = 0", "s = 0.05", "s = 0.75", "s = 2")
names(facet.labs) <- c(-0.75,-0.05,0,0.05,0.75,2)
p1<-freqs[q>0 & (s==-0.75 | s==-0.05 | s==0 | s==0.05 |  s==0.75 | s== 2)] %>% 
    ggplot(aes(x=q, fill=mutation)) + 
    geom_histogram(bins=21) +
    ggpubr::theme_pubr(base_size = 24) +
    facet_grid(s ~., scales = "free_y", labeller = labeller(s = facet.labs)) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    labs(subtitle="Frequencies [0,1]", x="Frequency")
p2<-freqs[q>0.05 & (s==-0.75 | s==-0.05 | s==0 | s==0.05 | s==0.75 | s== 2)] %>% 
    ggplot(aes(x=q, fill=mutation)) + 
    geom_histogram(bins=21) +
    ggpubr::theme_pubr(base_size = 24) +
    facet_grid(s ~., scales = "free_y", labeller = labeller(s = facet.labs)) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    labs(subtitle="Frequencies [0.05,1]", x="Frequency")
pmuts <- ggpubr::ggarrange(p1,p2, common.legend = TRUE, legend = "top" )
pmuts
```

```{r, fig.height=25, fig.width=12, fig.cap=caption}
caption <- "A. Relationship between selection coefficient (s) and pN/pS using different kinds of mutations defined by their frequency in the whole tumor. B. Distribution of the mutation frequencies when considering all the mutations in the tumor or just those present in at least 5% of the cells. Different row panels show different selection coefficient for the non-synomous (N) mutations."
#text.p <- ggpubr::ggparagraph(text = caption, face = "italic", size = 14, color = "black",)

ggpubr::ggarrange(p3, pmuts, nrow = 2,heights = c(5,16),labels = c("A","B")) 

```


See Table \@ref(tab:select-s)

```{r select-s}
dcast(t_pNpS_s,s~mutations) %>%
  knitr::kable(digits = 2,caption = "Average pN/pS for different s and threshold frequencies",booktabs=TRUE) %>%
  kableExtra::kable_styling(full_width = FALSE)

```


## Behaviour of pN/ps across time in the tumor progression

```{r}
freqs_time <- fread(here::here("results","time_freqs_merged.csv"))
#Limit to 3000 replicates,as some cases could have a couple more of them 
freqs_time <- freqs_time[, head(.SD,3000), by=V1002]
colnames(freqs_time) <- c("s", rep("N",500), rep("S",500),"id", rep("kk",4), "t")
freqs_time <- data.table::melt(freqs_time, id.vars = c("s","t"), measure.vars = 2:1001, variable.name = "mutation", value.name = "q")

```

```{r}
t_time_pNpS_s <- freqs_time[(s>-1),.(m0=sum(q>0),m01=sum(q>=0.001),m5=sum(q>=0.05),m100=sum(q==1)),by=.(s,t,mutation)] %>% 
  dcast(s + t ~mutation, value.var=c("m0","m01","m5","m100")) %>%
  .[,.(s, t = log10(t), pNpS_0 = m0_N/m0_S, pNpS_0.001 = m01_N/m01_S, pNpS_0.05 = m5_N/m5_S, pNpS_1 = m100_N/m100_S)] %>%
  data.table::melt(id.vars = c("s","t"), measure = patterns("^pNpS_"), variable.name="mutations", value.name="pN_pS") 
setattr(t_time_pNpS_s[["mutations"]],"levels",c("all", "q >= 0.001", "q >= 0.05", "clonal(q = 1)"))
```


```{r}
freq_thr <- 0.005
get_q_pNpS_table <- function(freq_thr){
  freqs_time[(s > 0.05), .(m = 500 * sum(q >= freq_thr) / .N), by = .(s, t, mutation)] %>%
    dcast(s + t ~ mutation, value.var = "m") %>%
    .[, .(s, t = log10(t), q = freq_thr, N, S, pNpS = N / S)] 
}

f <- c(1e-9,0.001, 0.005, 0.01, 0.02, 0.03, 0.04, 0.05, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1)
t_q_pNpS <- do.call( rbind, lapply(f, get_q_pNpS_table))
```


```{r fig.height=12, fig.width=10}

t_time_pNpS_s %>% 
  filter(mutations!="q >= 0.001", s > 0.01) %>%
  ggplot(aes(x=t, y=pN_pS, color=as.factor(s)))+
    #geom_smooth(method = "loess", lwd=2, se=FALSE)+
    geom_line(lwd=1.2) +
    geom_point() +
    ggpubr::theme_pubr(base_size=20)+
    labs(y="pN/pS", x="log10(cells)", color="      s") +
    geom_hline(yintercept = 1, lty="dashed") +
    facet_wrap(~mutations, ncol = 1, scales = "free_y") +
    #ylim(c(0,2)) +
    xlim(c(2,7)) 

```


```{r fig.height=8, fig.width=10}
anim <- t_q_pNpS %>% 
  ggplot(aes(x=q,y=pNpS)) +
    geom_point(aes(color=as.factor(s)), lwd=1.2) +
    ggpubr::theme_pubr(base_size=20)+
    labs(y="pN/pS", x="q", color="      s") +
    geom_hline(yintercept = 1, lty="dashed") +
    xlim(0,1) + 
    gganimate::transition_states(t, transition_length = 2, state_length = 3) +
    ggtitle('Tumor size: N =  10^{closest_state}')
gganimate::animate(anim, renderer = gganimate::ffmpeg_renderer(format = "webm"),width = 500, height = 500, res = 70)
```

```{r fig.height=16, fig.width=8, message=FALSE, warning=FALSE}
anim <- freqs_time [q > 0.05 & s >0.1] %>% 
  ggplot(aes(x=q, fill=mutation)) + 
  geom_histogram(bins=21) +
  ggpubr::theme_pubr(base_size = 20) +
  facet_grid(s ~., scales = "free_y") +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
  labs(subtitle="all mutations") +
  gganimate::transition_states(t, transition_length = 2, state_length = 3) +
  ggtitle('Tumor size: N =  {closest_state}')
gganimate::animate(anim, renderer = gganimate::ffmpeg_renderer(format = "webm"),width = 500, height = 800, res = 70)
  
```


# Discussion

[...]
 
# Aknowledgments

[...]
  
# References {-}
 
