---
title: "Figures using cohorts"
output:
  html_document:
    df_print: paged
  html_notebook:
    fig_caption: yes
    fig_height: 7
    fig_width: 10
    highlight: tango
    theme: readable
---
```{r setup, include=FALSE, message=FALSE}
library(tidyverse, quietly = T, warn.conflicts = FALSE)
library(data.table)
library(extrafont)
library(gridExtra)
library(latex2exp)
knitr::opts_chunk$set(message = FALSE, warning=FALSE, cache = FALSE,
                      fig.path="figures/", dev=c('png', 'pdf'),
                      echo = FALSE, fig.height = 12, fig.width = 16)
knitr::opts_knit$set(eval.after = "fig.cap")
caption <- "generic caption"
```

```{r load_data, include=FALSE}
csize <- 100 #max cohort size
files <- list.files(path = here::here("results"), pattern = "McF.*\\.result\\.csv", full.names = T)
big_table <- lapply(files, read_csv) %>% bind_rows() %>% filter(!is.na(s_pos), s_neg>=-1) %>% as.data.table
```

```{r}
big_table[,rep := 1:.N, by=c("s_pos","s_neg")]
big_table[,cohort := ceiling(rep/csize)]
```


```{r}
# Progression plot
prog_data <-big_table[rep<=3000 & s_pos>0,.(Time, tumor_size, n_clones, largest_clone),by=c("s_pos","s_neg","cohort")] 
mean_prog <- prog_data[,lapply(.SD, mean), by=c("s_pos")]
```




```{r fig-prog, fig.height=8, fig.width=16, fig.cap=caption, dev=c("png","pdf")}

fp1 <-prog_data %>%
  ggplot(aes(x=as.factor(s_pos),y=log10(Time), fill=as.factor(s_neg))) +
    geom_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggpubr::theme_pubr(base_size = 20)+
    ggsci::scale_fill_nejm() +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.47)) +
    theme(legend.position = "bottom") +
    labs(y="logartimic time", x="s+", fill="      s-", subtitle = "Progression time")
fp2 <-prog_data %>%
  ggplot(aes(x=as.factor(s_pos),y=n_clones, fill=as.factor(s_neg))) +
    geom_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggpubr::theme_pubr(base_size = 20)+
    ggsci::scale_fill_nejm() +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.47)) +
  
    theme(legend.position = "none") +
    ylim(c(0,10)) +
    labs(y="number of clones", x="s+", fill="      s-", subtitle = "Number of clones")
fp3 <-prog_data %>%
  ggplot(aes(x=as.factor(s_pos),y=largest_clone, fill=as.factor(s_neg))) +
    geom_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggsci::scale_fill_nejm() +
    ggpubr::theme_pubr(base_size = 20)+
    theme(axis.text.x = element_text(angle = 45, vjust = 0.47)) +
    theme(legend.position = "none") +
    ylim(c(0,1)) +
    labs(y="proportion of tumor size", x="s+", fill="      s-", subtitle = "Largest clone")


ggpubr::ggarrange(fp1,fp2,fp3, nrow=1, legend = "bottom", common.legend = TRUE)

caption <- "Parameters obtained at the end of the tumor progression for the different scenarios. Growth time represents the time, in decimal logarithmic scale, that the tumor took to reach the size of $10^7$ cells. The number of clones indicates the final number of clones present in the tumor, considering as clones the different genotypes with positive mutations present. The largest clone shows the proportion of the size of the tumor that occupies the majority clone."
```

```{r}
cohort_table <- big_table[rep<=3000 & s_pos>0, 
                          .(N_driver=sum(N_driver), S_driver=sum(S_driver),N_neutral=sum(N_neutral), S_neutral=sum(S_neutral),N_deletereous=sum(N_deletereous), S_deletereous=sum(S_deletereous),
                            sp_N_driver=sum(sp_N_driver), sp_S_driver=sum(sp_S_driver), sp_N_neutral=sum(sp_N_neutral), sp_S_neutral=sum(sp_S_neutral),sp_N_deletereous=sum(sp_N_deletereous), sp_S_deletereous=sum(sp_S_deletereous),
                            cl_N_driver=sum(cl_N_driver), cl_S_driver=sum(cl_S_driver),cl_N_neutral=sum(cl_N_neutral), cl_S_neutral=sum(cl_S_neutral),cl_N_deletereous=sum(cl_N_deletereous), cl_S_deletereous=sum(cl_S_deletereous)),by=c("s_pos","s_neg","cohort") ]
```

```{r}
cohort_table[,`:=`(pNpS_driver=N_driver/S_driver, pNpS_neutral=N_neutral/S_neutral, pNpS_deletereous=N_deletereous/S_deletereous)]
cohort_table[,`:=`(sp_pNpS_driver=sp_N_driver/sp_S_driver, sp_pNpS_neutral=sp_N_neutral/sp_S_neutral, sp_pNpS_deletereous=sp_N_deletereous/sp_S_deletereous)]
cohort_table[,`:=`(cl_pNpS_driver=cl_N_driver/cl_S_driver, cl_pNpS_neutral=cl_N_neutral/cl_S_neutral, cl_pNpS_deletereous=cl_N_deletereous/cl_S_deletereous)]
cohort_table[,lapply(.SD, mean), by=c("s_pos","s_neg")]
```


```{r}
pnps_table <- cohort_table[,.(pNpS_driver,pNpS_neutral,pNpS_deletereous,sp_pNpS_driver,sp_pNpS_neutral,sp_pNpS_deletereous,cl_pNpS_driver,cl_pNpS_neutral,cl_pNpS_deletereous), by=c("s_pos","s_neg")] %>% melt(id.vars = c("s_pos","s_neg"), value.name="pN_pS")
```




```{r fig-pnps, fig.height=15, fig.width=14, fig.cap=caption}
ind_plot <- function(X,y1=0,y2=2, tit){
  X %>%
   ggplot(aes(x=as.factor(s_pos),y=pN_pS, fill=as.factor(s_neg)))+
  stat_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggsci::scale_fill_nejm() +
 # stat_summary(fun.y=mean, geom = "line", aes(group=as.factor(s_pos)),size=0.8, alpha=0.7, position = position_dodge(width=0.75),show.legend = F, alpha=0.5) +
    #stat_summary(fun.y=mean, geom = "point",shape=23, size=1.2, aes(fill=as.factor(s_pos)), color="black", position = position_dodge(width=0.75),show.legend = F, alpha=0.5) +
      ggpubr::theme_pubr(base_size = 20)+
      geom_hline(yintercept = 1, lty="dashed") +
      labs(y="pN/pS", x="s+", fill="      s-", title=tit) +
     ylim(c(y1,y2))+
      theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
      theme(plot.title = element_text(face = "bold", size=18))
  
  
}

tit.lab=c(TeX('A. Driver gene, $pN/pS_{\\lbrack 0,1\\rbrack}$'),
                          TeX( 'B. Neutral gene, $pN/pS_{\\lbrack 0,1\\rbrack}$'),
                          TeX( 'C. Deletereous gene, $pN/pS_{\\lbrack 0,1\\rbrack}$'),
                          TeX( 'D. Driver gene, $pN/pS_{\\lbrack 0.05,1\\rbrack}$'),
                          TeX( 'E. Neutral gene, $pN/pS_{\\lbrack 0.05,1\\rbrack}$'),
                          TeX( 'F. Deletereous gene, $pN/pS_{\\lbrack 0.05,1\\rbrack}$'),
                          TeX( 'G. Driver gene, $pN/pS_{\\lbrack 1,1\\rbrack}$'),
                          TeX( 'H. Neutral gene, $pN/pS_{\\lbrack 1,1\\rbrack}$'),
                          TeX( 'I. Deletereous gene, $pN/pS_{\\lbrack 1,1\\rbrack}$'))

f1 <- pnps_table %>% filter(variable=="pNpS_driver") %>% ind_plot(y1=0,y2=2, tit.lab[1])
f2 <- pnps_table %>% filter(variable=="pNpS_neutral") %>% ind_plot(y1=0,y2=2, tit.lab[2])
f3 <- pnps_table %>% filter(variable=="pNpS_deletereous") %>% ind_plot(y1=0,y2=2, tit.lab[3])
f4 <- pnps_table %>% filter(variable=="sp_pNpS_driver") %>% ind_plot(y1=0,y2=60, tit.lab[4])
f5 <- pnps_table %>% filter(variable=="sp_pNpS_neutral") %>% ind_plot(y1=0,y2=2, tit.lab[5])
f6 <- pnps_table %>% filter(variable=="sp_pNpS_deletereous") %>% ind_plot(y1=0,y2=2, tit.lab[6])
f7 <- pnps_table %>% filter(variable=="cl_pNpS_driver") %>% ind_plot(y1=0,y2=60, tit.lab[7])
f8 <- pnps_table %>% filter(variable=="cl_pNpS_neutral") %>% ind_plot(y1=0,y2=2, tit.lab[8])
f9 <- pnps_table %>% filter(variable=="cl_pNpS_deletereous") %>% ind_plot(y1=0,y2=2, tit.lab[9])

ggpubr::ggarrange(f1,f2,f3, f4, f5,f6,f7,f8,f9, 
                  nrow = 3, ncol=3, common.legend = TRUE, heights = c(4,4,4)
                 )

caption <- "Average pN/pS, considering all the mutations present in the tumor, for each of the three genes (driver, neutral and deletereous) in all the scenarios simulated for the different positive ($s^+$, color lines) and negative ($s^-$, x-axis) selection coefficient. A.B,C show the pN/pS calculated using all the simulations present in the tumot. D, E, F using only those mutations above the threshold frequency of 0.05. G, H, I use only the clonal mutations for the tumor, those present in all the cells.
"
```

## VAF

```{r load-freqs}
freqsQ <- fread(here::here("results","s_freqs_merged.csv"))
```

```{r}
# Temp.
files2 <- list.files(path = here::here("results"), pattern = "McF_31\\.freqs\\.csv", full.names = T)
freqs <- lapply(files2, read_csv, col_names=FALSE) %>% bind_rows()
```

```{r}
q <- freqs[1:100,1:3000] %>% as.matrix %>% as.numeric
nq <- 100
tFreqs <- data.frame(q=q, 
                      gene=factor(c(rep("driver",nq*1000),rep("deletereous",nq*1000),rep("neutral",nq*1000)), 
                                 levels = c("driver","neutral","deletereous")),
                      mutation=factor(c(rep("N",nq*500),rep("S",nq*500),rep("N",nq*500),rep("S",nq*500),rep("N",nq*500),rep("S",nq*500)), 
                                 levels = c("N","S")))
```

```{r}
examplePNPS <- as.data.table(tFreqs)[q>0,.N, by=c("gene","mutation")] %>% 
                              dcast(gene ~ mutation) %>% 
                              mutate(pNpSall=N/S) %>% 
                              select(gene,pNpSall) %>%
                left_join( as.data.table(tFreqs)[q>0.05,.N, by=c("gene","mutation")] %>%
                             dcast(gene ~ mutation) %>% 
                             mutate(pNpSdet=N/S) %>% 
                             select(gene,pNpSdet) ) %>%
                left_join( as.data.table(tFreqs)[q==1,.N, by=c("gene","mutation")] %>%
                             dcast(gene ~ mutation) %>% 
                             mutate(pNpSclon=N/S) %>% 
                             select(gene,pNpSclon) ) 
examplePNPS
```


```{r fig-VAF-example, fig.height=6, fig.width=12, fig.cap = caption}
tFreqs %>% 
  filter(q>0) %>% 
  ggplot(aes(x=q, fill=mutation)) + 
    #geom_histogram(bins=11) +
  geom_bar(position = "stack") +
    scale_x_binned(breaks = seq(0.05,0.95, by = 0.05),
                   labels = c(".05","",".15","",".25","",".35","",".45","",".55","",".65","",".75","",".85","",".95"))+
    
    facet_wrap(~gene) +
  
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=30000, fill=NULL, 
              label=TeX('$pN/pS_{\\lbrack 0,1\\rbrack}$ = ', output = "character")),
              vjust = 1, hjust=1, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=30000, fill=NULL, 
              label=sprintf(" %.02f",pNpSall)),
              vjust = 1, hjust=0, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=26000, fill=NULL, 
              label=TeX('$pN/pS_{\\lbrack 0.05,1\\rbrack}$ = ', output = "character")),
              vjust = 1, hjust=1, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=26000, fill=NULL, 
              label=sprintf(" %.02f",pNpSdet)),
              vjust = 1, hjust=0, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=22000, fill=NULL, 
              label=TeX('$pN/pS_{\\lbrack 1,1\\rbrack}$ = ', output = "character")),
              vjust = 1, hjust=1, size=5, parse=T) +
  geom_text(data = examplePNPS, mapping = aes(x=0.8, y=22000, fill=NULL, 
              label=sprintf(" %.02f",pNpSclon)),
              vjust = 1, hjust=0, size=5, parse=T) +
  
    labs(x="Frequency", y="ocurrences in cohort", color="mutation type", title=latex2exp::latex2exp("$s^+ = 0.1$, $s^- = -0.25$, 1 cohort 100 tumors")) +
    ggpubr::theme_pubr(base_size = 18) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1))
caption <- "Variant allele in a cohort of 1000 tumors for the $s^+$ = 0.1 and $s^-$ = -0.25 scenario."
```

## Relationship between selection coefficient and pN/pS


```{r}
freqs <- fread(here::here("results","s_freqs_merged.csv"))
#Limit to 3000 replicates,as some cases could have a couple more of them 
freqs <- freqs[, head(.SD,3000), by=V1002]
colnames(freqs) <- c("s", rep("N",500), rep("S",500),"id", rep("kk",4))
freqs <- data.table::melt(freqs, id.vars = "s", measure.vars = 2:1001, variable.name = "mutation", value.name = "q")
```

```{r}
t_pNpS_s <- freqs[(s>-1),.(m0=sum(q>0),m01=sum(q>=0.001),m5=sum(q>=0.05),m100=sum(q==1)),by=.(s,mutation)] %>% 
  dcast(s~mutation, value.var=c("m0","m01","m5","m100")) %>%
  .[,.(s, pNpS_0 = m0_N/m0_S, pNpS_0.001 = m01_N/m01_S, pNpS_0.05 = m5_N/m5_S, pNpS_1 = m100_N/m100_S)] %>%
  data.table::melt(id.vars = "s", measure = patterns("^pNpS_"), variable.name="mutations", value.name="pN_pS") 
setattr(t_pNpS_s[["mutations"]],"levels",c("all", "q >= 0.001", "q >= 0.05", "clonal(q = 1)"))
```


```{r fig.height=6, fig.width=12}
 p3 <- t_pNpS_s %>% 
  filter(mutations!="q >= 0.001") %>%
  ggplot(aes(x=s, y=pN_pS, color=mutations))+
  
    #geom_smooth(method = "loess", lwd=2, se=FALSE)+
    geom_line(lwd=1.2) +
    geom_point() +
     ggpubr::theme_pubr(base_size=20)+
  
    geom_hline(yintercept = 1, lty="dashed") +
    geom_vline(xintercept = 0, lty="dashed") +
    ggsci::scale_color_aaas(name = "", labels = unname(TeX(c("$pN/pS_{\\[0,1\\]}$", 
                                               "$pN/pS_{\\[0.05,1\\]}$",
                                               "$pN/pS_{\\[1,1\\]"))))+
  #ggsci::scale_color_nejm() +
    labs(x="s", y="pN/pS") +
    xlim(c(-1,2)) 
```


```{r}
facet.labs <- c("s = -0.75","s = -0.05","s = 0", "s = 0.05", "s = 0.75", "s = 2")
names(facet.labs) <- c(-0.75,-0.05,0,0.05,0.75,2)
p1<-freqs[q>0 & (s==-0.75 | s==-0.05 | s==0 | s==0.05 |  s==0.75 | s== 2)] %>% 
    ggplot(aes(x=q, fill=mutation)) + 
    geom_bar(position = "stack") +
    scale_x_binned(breaks = seq(0.05,0.95, by = 0.05),
                   labels = c(".05","",".15","",".25","",".35","",".45","",".55","",".65","",".75","",".85","",".95"))+
    ggpubr::theme_pubr(base_size = 20) +
    facet_grid(s ~., scales = "free_y", labeller = labeller(s = facet.labs)) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    labs(subtitle="Frequencies [0,1]", x="Frequency")
p2<-freqs[q>0.05 & (s==-0.75 | s==-0.05 | s==0 | s==0.05 | s==0.75 | s== 2)] %>% 
    ggplot(aes(x=q, fill=mutation)) + 
    geom_bar(position = "stack") +
    scale_x_binned(breaks = seq(0.1,0.95, by = 0.05),
                   labels = c("",".15","",".25","",".35","",".45","",".55","",".65","",".75","",".85","",".95"))+
    ggpubr::theme_pubr(base_size = 20) +
    facet_grid(s ~., scales = "free_y", labeller = labeller(s = facet.labs)) +
    theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust = 1)) +
    labs(subtitle="Frequencies [0.05,1]", x="Frequency")
pmuts <- ggpubr::ggarrange(p1,p2, common.legend = TRUE, legend = "top" )
```

```{r fig-selection, fig.height=25, fig.width=12, fig.cap=caption}
caption <- "A. Relationship between selection coefficient (s) and pN/pS using different kinds of mutations defined by their frequency in the whole tumor. B. Distribution of the mutation frequencies when considering all the mutations in the tumor or just those present in at least 5% of the cells. Different row panels show different selection coefficient for the non-synomous (N) mutations."
ggpubr::ggarrange(p3, pmuts, nrow = 2,heights = c(5,16),labels = c("A","B")) 
```




```{r select-s}
dcast(t_pNpS_s,s~mutations) %>%
  knitr::kable(digits = 2,caption = "Average pN/pS for different s and threshold frequencies",booktabs=TRUE) %>%
  kableExtra::kable_styling(full_width = FALSE)

```

# Time

```{r}
# Get the new data in files time_s_pos_XX_pY.{result|freqs}.csv and combine them
# in a big data.table
freq_files <- list.files(path = here::here("results"), pattern='time_s_pos_.*_p.*freqs',full.names = TRUE)
res_files <- list.files(path = here::here("results"), pattern='time_s_pos_.*_p.*result',full.names = TRUE)

hrep <- str_extract(res_files,'_p[0-9]*\\.') %>% parse_number()
dt_res <- rbindlist(setNames(lapply(res_files, fread, select = c(29,35 ,34, 23, 27), col.names = c("s", "rep","threshold_size","true_size","time") ), hrep), idcol = "hrep")
hrep <- str_extract(res_files,'_p[0-9]*\\.') %>% parse_number()
dt_freq <- rbindlist(setNames(lapply(freq_files, fread, col.names = c(rep("N",500), rep("S",500),"id", "s", rep("kk",4), "threshold_size", "rep")), hrep), idcol = "hrep")

# Combine datatables
full_data <- dt_res[dt_freq, on= c("s","hrep","rep","threshold_size")]

#Clean data
full_data <- full_data[,-c("id", "kk", "kk.1", "kk.2", "kk.3")]
full_data[,threshold_size := log10(threshold_size)]
full_data[,true_size := log10(true_size)]
full_data[,rep := as.numeric(factor(paste(hrep,rep,sep="_"))),by=s]
rm(list = c("dt_freq","dt_res"))
```

```{r}
# Number of replicates
 full_data[,.N,by=c("s","threshold_size")] %>% dcast(s ~ threshold_size, value.var="N")
```


```{r}
cohort_size <- 100
time_points <- 6 
# Remove incomplete replicates
wdt <-  full_data[threshold_size<=(time_points+1)][,full_rep := .N>=time_points, by=c("s","rep")]
rm(full_data)
wdt[, rel_time := time - .SD[1,time], by=c("s","rep")]
wdt <- wdt[which(full_rep==TRUE)]

# Some replicates can fail in the last point, remove them
if(time_points == 7){
  wdt[, wrong := (.SD[7,time] - .SD[6,time] < 0), by=c("s","rep")]
  wdt <- wdt[which(wrong==FALSE)]
} else{
  wdt[,wrong := FALSE]
}


# Renumber valid replicates
wdt[, rep := 1:.N, by=c("s","threshold_size")]
setcolorder(wdt, c("s","rep"))

# Limit to 10K replicates
wdt <- wdt[rep<=10000]
# Create cohorts (100 tumors)
wdt[,cohort := ceiling(rep/cohort_size)]

# discard incomplete cohorts
wdt[, full_cohort := .N==cohort_size, by=c("s","cohort", "threshold_size")]
wdt <- wdt[full_cohort == TRUE,][,c("full_cohort","hrep","full_rep","wrong") := NULL]


#Calculate relative time (relative to N=10^2)
setcolorder(wdt, c("s","cohort", "rep", "threshold_size", "true_size", "time", "rel_time"))

```

```{r}
wdt[,.(cohorts=max(cohort)),by=c("s")] 
```



```{r}
#clonal, replicate by replicate, then summarizing by cohort
clonal <- wdt[,.(s, cohort, rep, threshold_size, true_size, time, rel_time, Nmuts= apply(wdt[,which(grepl("^N.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X==1)}), Smuts= apply(wdt[,which(grepl("^S.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X==1)}))][, .(true_size=mean(true_size), rel_time=mean(rel_time, na.rm = T), Nmuts = mean(Nmuts), Smuts = mean(Smuts)), by = c("s","cohort","threshold_size")][,pNpS := Nmuts/Smuts]

```

```{r}
#clonal, replicate by replicate, then summarizing by cohort
segregating <- wdt[,.(s, cohort, rep, threshold_size, true_size, time, rel_time, Nmuts= apply(wdt[,which(grepl("^N.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X>=0.05)}), Smuts= apply(wdt[,which(grepl("^S.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X>=0.05)}))][, .(true_size=mean(true_size), rel_time=mean(rel_time, na.rm = T), Nmuts = mean(Nmuts), Smuts = mean(Smuts)), by = c("s","cohort","threshold_size")][,pNpS := Nmuts/Smuts]

```

```{r}
#clonal, replicate by replicate, then summarizing by cohort
all <- wdt[,.(s, cohort, rep, threshold_size, true_size, time, rel_time, Nmuts= apply(wdt[,which(grepl("^N.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X>0)}), Smuts= apply(wdt[,which(grepl("^S.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X>0)}))][, .(true_size=mean(true_size), rel_time=mean(rel_time, na.rm = T), Nmuts = mean(Nmuts), Smuts = mean(Smuts)), by = c("s","cohort","threshold_size")][,pNpS := Nmuts/Smuts]

```

```{r}
panA <- all %>%
  ggplot(aes(x=as.factor(threshold_size), y=Nmuts, color=as.factor(s))) + 
     geom_point(size=0.5,alpha=0.5) +
    stat_summary(fun.y=mean, geom = "line", aes(group=as.factor(s)),size=0.8) +
    stat_summary(fun.y=mean, geom = "point", aes(group=as.factor(s)), size=1) +
    ggsci::scale_color_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="N", x=TeX("log_{10}(cells)"), color="      s", title= TeX("\\textbf{A.} $N_{\\lbrack 0,1\\rbrack}$")) +
    theme(plot.title = element_text(face = "bold", size=14))
  
panB <- all %>%
  ggplot(aes(x=as.factor(threshold_size), y=Smuts, color=as.factor(s))) + 
     geom_point(size=0.5,alpha=0.5) +
    stat_summary(fun.y=mean, geom = "line", aes(group=as.factor(s)),size=0.8) +
    stat_summary(fun.y=mean, geom = "point", aes(group=as.factor(s)), size=1) +
    ggsci::scale_color_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none", ) +
    labs(y="S", x=TeX("log_{10}(cells)"), color="      s", title= TeX("\\textbf{B.} $S_{\\lbrack 0,1\\rbrack}$"))  +
    theme(plot.title = element_text(face = "bold", size=14))

panC <- all %>%
  ggplot(aes(x=as.factor(threshold_size), y=pNpS, fill=as.factor(s))) + 
    # geom_point(size=0.5,alpha=0.5) +
     stat_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggsci::scale_fill_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="pN/pS", x=TeX("log_{10}(cells)"), fill="      s", title= TeX("\\textbf{C.} $pN/pS_{\\lbrack 0,1\\rbrack}$"))  +
    geom_hline(yintercept = 1, lty="dashed") +
    ylim(c(0,50)) +
    theme(plot.title = element_text(face = "bold", size=14))

panD <- segregating %>%
  ggplot(aes(x=as.factor(threshold_size), y=Nmuts, color=as.factor(s))) + 
     geom_point(size=0.5,alpha=0.5) +
    stat_summary(fun.y=mean, geom = "line", aes(group=as.factor(s)),size=0.8) +
    stat_summary(fun.y=mean, geom = "point", aes(group=as.factor(s)), size=1) +
    ggsci::scale_color_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="N", x=TeX("log_{10}(cells)"), color="      s", title= TeX("\\textbf{D.} $N_{\\lbrack 0.05,1\\rbrack}$"))  +
    theme(plot.title = element_text(face = "bold", size=14))
  
panE <- segregating %>%
  ggplot(aes(x=as.factor(threshold_size), y=Smuts, color=as.factor(s))) + 
     geom_point(size=0.5,alpha=0.5) +
    stat_summary(fun.y=mean, geom = "line", aes(group=as.factor(s)),size=0.8) +
    stat_summary(fun.y=mean, geom = "point", aes(group=as.factor(s)), size=1) +
    ggsci::scale_color_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="S", x=TeX("log_{10}(cells)"), color="      s", title= TeX("\\textbf{E.} $S_{\\lbrack 0.05,1\\rbrack}$")) +
    theme(plot.title = element_text(face = "bold", size=14))

panF <- segregating %>%
  ggplot(aes(x=as.factor(threshold_size), y=pNpS, fill=as.factor(s))) + 
     stat_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggsci::scale_fill_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="pN/pS", x=TeX("log_{10}(cells)"), fill="      s", title= TeX("\\textbf{F.} $pN/pS_{\\lbrack 0.05,1\\rbrack}$")) +
    geom_hline(yintercept = 1, lty="dashed") +
    ylim(c(0,100)) +
    theme(plot.title = element_text(face = "bold", size=14))

panG <- clonal %>%
  ggplot(aes(x=as.factor(threshold_size), y=Nmuts, color=as.factor(s))) + 
     geom_point(size=0.5,alpha=0.5) +
    stat_summary(fun.y=mean, geom = "line", aes(group=as.factor(s)),size=0.8) +
    stat_summary(fun.y=mean, geom = "point", aes(group=as.factor(s)), size=1) +
    ggsci::scale_color_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="N", x=TeX("log_{10}(cells)"), color="      s", title= TeX("\\textbf{G.} $N_{\\lbrack 1,1\\rbrack}$")) +
    theme(plot.title = element_text(face = "bold", size=14))
  
panH <- clonal %>%
  ggplot(aes(x=as.factor(threshold_size), y=Smuts, color=as.factor(s))) + 
     geom_point(size=0.5,alpha=0.5) +
    stat_summary(fun.y=mean, geom = "line", aes(group=as.factor(s)),size=0.8) +
    stat_summary(fun.y=mean, geom = "point", aes(group=as.factor(s)), size=1) +
    ggsci::scale_color_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="S", x=TeX("log_{10}(cells)"), color="      s", title= TeX("\\textbf{H.} $S_{\\lbrack 1,1\\rbrack}$")) +
    theme(plot.title = element_text(face = "bold", size=14))

panI <- clonal %>%
  ggplot(aes(x=as.factor(threshold_size), y=pNpS, fill=as.factor(s))) + 
     stat_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggsci::scale_fill_nejm() +
    ggpubr::theme_pubr(base_size=16, legend = "none") +
    labs(y="pN/pS", x=TeX("log_{10}(cells)"), fill="      s", title= TeX("\\textbf{I.} $pN/pS_{\\lbrack 1,1\\rbrack}$")) +
    geom_hline(yintercept = 1, lty="dashed") +
    ylim(c(0,50)) +
    theme(plot.title = element_text(face = "bold", size=14))



```

```{r fig-time, fig.height=13, fig.width=12}

ggpubr::ggarrange(panA,panB,panC,panD,panE,panF,panG,panH,panI, ncol=3, nrow=3, widths = c(2,2,6), common.legend = T,align = "hv")
```

# McDonald-Kreitman

```{r}
#clonal, replicate by replicate, then summarizing by cohort
mutations_table <- wdt[,.(s, cohort, rep, threshold_size,  
                          Npol= apply(wdt[,which(grepl("^N.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X>0 & X<1)}), 
                          Spol= apply(wdt[,which(grepl("^S.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X>0 & X<1)}),
                          Nfix= apply(wdt[,which(grepl("^N.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X==1)}), 
                          Sfix= apply(wdt[,which(grepl("^S.*$", colnames(wdt))), with=FALSE], 1, function(X){sum(X==1)}))
                       ][, .(Nfix = sum(Nfix), Sfix = sum(Sfix),Npol = sum(Npol), Spol = sum(Spol)), by = c("s","cohort","threshold_size")
                        ][,pNpS := Npol/Spol
                        ][,dNdS := Nfix/Sfix
                        ][,alpha := 1-pNpS/dNdS
                        ]
mutations_table[,pvalue := apply(mutations_table[,.(Npol,Spol,Nfix,Sfix)],1,function(X) fisher.test(matrix(X,nr=2,byrow = T))$p.value)]

```


```{r fig-mcDonald, fig.height=6,fig.width=10, dev=c('png','pdf')}
mutations_table %>%
  ggplot(aes(x=as.factor(threshold_size), y=alpha, fill=as.factor(s))) + 
     stat_boxplot(outlier.size=1, outlier.color = NA, outlier.shape = 21) +
    ggsci::scale_fill_nejm() +
    ggpubr::theme_pubr(base_size=20) +
    labs(y=expression(alpha), x=latex2exp::TeX("log_{10}(cells)"), fill="      s")  +
    geom_hline(yintercept = 1, lty="dashed") +
    geom_hline(yintercept = 0, lty="dashed") +
    ylim(c(-15,1)) +
    theme(plot.title = element_text(face = "bold", size=14))

```



