---
title: 'Evolution of the pN/pS in tumors: scenarios with different amount of loci under positive and negative selection under the McFarland growth model'
date: "25/12/2017"
author: "Andrés Pérez-Figueroa"
output:
  html_document:
    fig_caption: yes
    fig_height: 7
    fig_width: 10
    highlight: tango
    theme: readable
  pdf_document: 
    fig_caption: yes
    fig_height: 7
    fig_width: 10
  html_notebook:
    fig_caption: yes
    fig_height: 7
    fig_width: 10
    highlight: tango
    theme: readable
---
```{r setup, include=FALSE, message=FALSE}
library(tidyverse, quietly = T, warn.conflicts = FALSE)
library(extrafont)
knitr::opts_chunk$set(message = FALSE, warning=FALSE, cache = TRUE, 
                      echo = FALSE, fig.height = 12, fig.width = 16)
knitr::opts_knit$set(eval.after = "fig.cap")
caption <- "generic caption"
```

This report documents the results of the run for different scenarios in the simulations using *OncoSimulR* (Díaz-Uriarte 2017). Here we used the McFarland et al. (2013) growth model, where the effects of drivers (non-synonymous mutations with positive effect) contribute to the numerator of the birth rate, and those of the deleterious (NS mutations with negative effect) passengers to the denominator as:
$$b = \frac{(1+s^+)^{n_{drivers}}}{(1-s^-)^{n_{deleterious}}}$$
For death rate, *oncoSimulR* uses the expression that McFarland et al. (2013) for large cancers (grown to $10^6$ cells)”: $$d = \log(1+N/K)$$
where $N$ here is the current tumor size and $K$ is the initial equilibrium population size. As the authors explain, for large $N/K$ the above expression recapitulates Gompertzian dynamics observed experimentally for large tumors. Under the current neutral scenario, this means that population size would be buffered by the changes in the death rate.

We arbitrarily created a genome of 10000 synonymous sites (S). The number of non-synonymous sites (N) used is 27600, to fit the estimation of $N/S = 2.76$ obtained in human soma based on codon usage and mutational spectrum (Milhollland et al. 2017). This yield a total number of 37600 sites. The mutation rate is set in $2.66\times10^{-9}$ as representative of human somatic mutation (Milholland et al. 2017). A starting population of 100000 cells, all with the same genotype, is evolved for 1000 time units, taking samples every 5 time units (so, 201 samples taken). For each sample we obtain both all the variants present in the population or just those above a threshold frequency of 0.05. This simulation setting is replicated 100 times.
Different scenarios (Table 1) have different distribution of the N sites regarding its selective effect (positive: s=0.1; negative: s=-0.1; neutral: s=0). The parameters used were chosen by trial and error. We have varied the value for number of nonsynonymous loci with positive selection until finding values that allow us to run the simulations in a reasonable time, with a manageable amount of results and that we would ensure that tumors were obtained that would make the population grow by (at least) an order of magnitude.

    

```{r read_data}
samples <- readRDS(here::here("results","sel_All_McFL.data.rds"))
colorcase <- c("black","darkorange","blue2","purple","darkgreen","red")
colortype <- c("N"="#D55E00","S"="#56B4E9")
```

```{r preprocess_data}
sce_labels <- c("neutral","symmetric","10x negative", "mostly negative", "all positive", "all negative")
levels(samples$case) <- sce_labels

## ADD replicate variable
reps <- samples %>% group_by(case) %>% filter(d==1e-09,t==0) %>% count
times <- samples %>% group_by(case) %>% count
loop <- times$n / reps$n
samples$rep <- rep(purrr::map(reps$n,seq) %>% unlist, rep(loop,reps$n))
params <- tibble::tibble(
  scenario = sce_labels,
  N_pos = c(0,500,500,500,500,0),
  N_neg = c(0,500,5000,27000,0,27000),
  N_neu = c(27600,26600,22100,100,27100,100),
  replicates = reps$n
)

```

```{r replicates_table, results='asis'}
knitr::kable(params, 
             caption = "Parameters used in each scenario. The number of replicates is the obtained after some of them failed",
             col.names = c("scenario", "N positive", "N negative","N neutral", "replicates") 
             )
```


## Results

```{r, fig.width=10, fig.height=8, fig.cap=caption}
caption <- "Population growth under the symetric scenario measured as the log10-scaled number of cells by time unit. Grey lines show the  individual replicates. Red line represents the mean across replicates"

p <- samples %>% filter(d==1e-9) %>% ggplot(aes(x=t,y=log10(TS))) + 
  geom_line(aes(group=rep), color="gray20", cex=0.8, alpha=0.1) + 
  #geom_smooth(color="red",se=F) + 
  stat_summary(fun.y = mean, geom = "line", color="red", cex=1.06) +
  
  ylim(5,7.5) +
  
  facet_wrap(~case, nrow=2) +
  theme_bw(base_size = 18) +
  theme(panel.grid = element_blank()) +
  labs(y=expression(paste(log[10],"(cells)")),
       x= "t"
      )
  
print(p)

```



```{r, fig.width=10, fig.height=7, fig.cap=caption}
caption <- "Evolution of the pN/pS under the symetric scenario.Grey points show the values for the individual replicates in each sampled time unit. Red line represents the smoothed average across replicates. Dashed black line show the reference value pN/pS=1."

p <- samples %>% filter(d==1e-9) %>% ggplot(aes(x=t,y=pN_pS)) + 
  geom_point(aes(group=rep), shape=21, color="gray20", fill="gray80", alpha=0.2, cex=0.3) + 
  #geom_smooth(color="red",se=F) + 
  stat_summary(fun.y = mean, geom = "line", color="red", cex=1.06) +
  geom_abline(slope = 0, intercept = 1, lty = "dashed" ) + 
  ylim(0,5) +
  
  facet_wrap(~case, nrow=2) +
  theme_bw(base_size = 18) + theme(panel.grid = element_blank()) + 
  ylab("pN/pS") + xlab("t") 
  
print(p)

```





```{r, fig.width=10, fig.height=7, fig.cap=caption}
caption <- "Evolution of the pN/pS parameter under the different scenarios simulated. Solid lines represent the smoothed average across replicates for each scenario. Dashed black line show the reference value pN/pS=1."
data <- samples %>% filter(d==1e-9) 
data$case <- factor(data$case, levels=levels(data$case)[c(2,5,1,3,4,6)])
p <- data %>% ggplot() +
  geom_smooth(aes(x=t,y=pN_pS, color=case), se=FALSE, cex=2) +
  geom_abline(slope = 0, intercept = 1, lty = "dashed" ) +
  theme_classic(base_size = 28,base_family = "sans") + 
  scale_color_manual("",
                     values= colorcase[c(2,5,1,3,4,6)],
                     guide = guide_legend(direction = "vertical")) +
  theme(legend.position = c(0.8,0.9)) +
  scale_y_continuous(limits = c(0,2), expand = c(0,0)) +
  scale_x_continuous(limits = c(0,1050), expand =  c(0,0)) +
  ylab("pN/pS") 
print(p)
```

```{r, eval= FALSE}
  grDevices::cairo_pdf(here::here("FigpNpS.pdf"),width = 10)
print(p)
dev.off()

```



```{r, fig.width=10, fig.height=7, fig.cap=caption}
caption <- "Population growth under the different scenarios measured as the number of cells by time unit. Solid lines represent the smoothed average across replicates for each scenario."
data <- samples %>% filter(d==1e-9) 
data$case <- factor(data$case, levels=levels(data$case)[c(5,4,2,3,1,6)])
p <- data %>% ggplot() +
  geom_smooth(aes(x=t,y=log10(TS), color=case), se=FALSE, cex=2) +
  theme_classic(base_size = 28,base_family = "sans") + 
  scale_color_manual("",
                     values= colorcase[c(5,4,2,3,1,6)],
                     guide = guide_legend(direction = "vertical")) +
  theme(legend.position = c(0.3,0.9)) +
  scale_y_continuous(expand = c(0,0.05)) +
  scale_x_continuous(limits = c(0,1050), expand =  c(0,0)) +
  labs(y=expression(paste(log[10],"(cells)")),
       x= "t"
      ) 
print(p)
```

```{r, eval=FALSE}
  grDevices::cairo_pdf(here::here("FigTS.pdf"),width = 10)
print(p)
dev.off()

```


```{r}
pNpS_summary <- samples  %>% group_by(case,d, t) %>%  
  summarise(
    TS=mean(TS, na.rm=T),
    mean_pN_pS=mean(pN_pS, na.rm = T),
    median_pN_pS=median(pN_pS, na.rm = T),
    N=mean(N, na.rm = T),
    S = mean(S, na.rm = TRUE)
    )
```

```{r, fig.width=10, fig.height=7, fig.cap=caption}
caption <- "Accumulation of synonymous and nonsynonymous mutations across the time for the different scenarios simulated averaged by replicates. Blue areas represent the number of synonymous mutations, and red areas represent the number of nonsynonimous mutations stacked to the the number of synonymous. The total colored area represent then the total number of mutations"
mutations <- pNpS_summary %>% filter(d==1e-9) %>% 
  gather(key = "Type", value = "Mean", factor_key = T, N,S) %>% 
  select(t,TS,d,Type,Mean, case) %>% filter(d==1e-9)

p <- ggplot(mutations) + 
  geom_area(aes(x=t,y=Mean, fill=Type), position="stack") +
  facet_wrap(~case, nrow = 2) +
  scale_fill_manual(values = colortype) +
  theme_bw(base_size = 18) + ylab("Average number of mutations") + xlab("t") +
  theme(legend.position = "top", legend.text = element_text(size = 12)) +
  theme(legend.title = element_text(size = 12) ) + labs(fill="Mutation type:   ")
print(p)
```

```{r, fig.width=9, fig.height=9, fig.cap=caption}
caption <- "Distribution of the number of synonymous (S) and nonsynonymous (N) mutations across replicates at the end of the simulations under the different scenarios. Colored areas represent the kernel density estimation, with bandwidth = 0.1, for the number of mutations of an specific type. Points represent the mean of the distribution. Red colors stand for nonsynonimous mutations while blue stand forsynonymous. Scenarios in the y-axis are oredered by increasing average of nonsynonymous mutations."
data <- samples %>% filter(d==1e-9,t==1000) 
mean_muts <- data %>% group_by(case) %>% 
  select(case, N,S) %>% summarise_all(median) %>% arrange(N)
data$case <- factor(data$case, levels(data$case)[mean_muts$case])
colorcase2 <- colorcase[mean_muts$case]
mean_muts <- mean_muts %>% 
  gather(key = "Type", value = "Mean", factor_key = T, S, N) 
data %>% 
  gather(key = "Type", value = "Count", factor_key = T, S, N) %>% 
  ggplot() +
  ggridges::geom_density_ridges(aes(x=log10(Count), y=case, fill=Type, alpha=0.65), 
                                alpha=0.5, scale=0.92, bandwidth=0.1) +
  scale_fill_manual(values = colortype) +
  scale_color_manual(values = colortype) +
  #geom_path(data=mean_muts, aes(x=log10(Mean), y=case, group=Type, color=Type),cex=1.3) +
  geom_point(data=mean_muts, aes(x=log10(Mean), y=case, fill=Type), shape=21, color="black", cex=1.5) +
  #scale_color_manual("mutation",values = c("darkblue","darkred")) +
  theme_bw(base_size = 18) +
  theme(legend.position = "top", legend.text = element_text(size = 12)) +
  theme(legend.title = element_text(size = 12) ) + 
  labs(x=expression(paste(log[10],"(count)")),
       y= "scenario",
       color = "Mutation type:  ",
       fill = "Mutation type:  "
      )

```

```{r fig_pN_pS, fig.width=10, fig.height=7, fig.cap=caption}
caption <- "Relationship between the proportion of nonsynonymous mutated sites (pN) and the proportion of synonymous mutated sites (pS). Background gradient represent the pN/pS ratio space. Solid lines represent the smoothed regression for pN over pS under the different scenarios. Dashed diagonal is the reference for pN/pS = 1, while the dotted lines represent the reference for pN/pS values between 0.1 to 2 with 0.1 increase."
data <- pNpS_summary
data$case <- factor(pNpS_summary$case, levels(pNpS_summary$case)[c(5,2,3,1,4,6)])
colorcase2 <- colorcase[c(5,2,3,1,4,6)]
col_scale <- rev(terrain.colors(21))

x <- seq(0.0000001,0.03, length.out = 300)
z <- expand.grid(x,x)  
z$pN_pS <- z[,2]/z[,1]
z$pN_pS[z$pN_pS>=2] <- 2

p <- data %>% ggplot(aes(x=(S/10000),y=(N/27600))) + 
  geom_raster(data=z, aes(Var1,Var2,fill=pN_pS)) +
  scale_fill_gradientn(colours = col_scale) +
  theme(legend.title = element_text(size = 12) ) + 
  geom_abline(slope=1,intercept = 0, lty="dashed", cex=0.5, alpha=0.5) +
  geom_abline(slope=seq(0.1,2,by=0.1),intercept = 0, lty="dotted", cex=0.05) +
    #geom_point(aes(color=case), cex=1, alpha=0.2) +
  geom_smooth(aes(color=case), se=F, cex=1.5) +
  scale_color_manual(values = colorcase2) +
  theme_classic(base_size=18) +
  scale_y_continuous(expand = c(0,0)) +
  scale_x_continuous(expand = c(0,0)) +
  labs(fill="pN/pS", 
       color="scenario",
       x="pS",
       y="pN"
       ) 
print(p)
```


### Effect of the detection threshold and clonal mutations

In general, almost all mutations were found at very low frequency, except for some drivers. Some of these (after visual inspection in some replicas) become clonal (fixed) but generally no more than 1 or 2. Therefore, when sampling those mutations with a frequency higher than 0.05 we find almost no mutations, especially there were no synonymous mutations above the threshold (Table 2). This may be due to the fact that our tumor, given the simulated parameters, is "not very aggressive" and most drivers appear independently, competing with each other.
```{r  results='asis'}
samples %>% filter(d==0.05, S>0, t==1000) %>% group_by(case) %>% 
  summarise(N=mean(N), S=mean(S), rep=n()) %>%
  knitr::kable( digits = 1,
             caption = "Average number of nonsynonymous and synonymous mutations with frequency above 0.05 on those replicates with at least one synonymous mutation over that threshold",
             col.names = c("scenario", "N", "S", "replicates") 
             )
```


## References

Diaz-Uriarte, R. (2017). OncoSimulR: Genetic simulation with arbitrary epistasis and mutator genes in asexual populations. Bioinformatics, 33(12), 1898–1899. http://doi.org/10.1093/bioinformatics/btx077

McFarland, C. D., Korolev, K. S., Kryukov, G. V., Sunyaev, S. R., & Mirny, L. A. (2013). Impact of deleterious passenger mutations on cancer progression. Proceedings of the National Academy of Sciences, 110(8), 2910–2915. http://doi.org/10.1073/pnas.1213968110

Milholland, B., Dong, X., Zhang, L., Hao, X., Suh, Y., & Vijg, J. (2017). Differences between germline and somatic mutation rates in humans and mice. Nature Communications, 8(May), 1–8. http://doi.org/10.1038/ncomms15183






